src/state/persist.rs

```diff
@@ -0,0 +1,136 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_hamt::Hamt;
+use microkelvin::{HostStore, Ident, OffsetLen, StoreRef};
+use rkyv::ser::{serializers::AllocSerializer, Serializer};
+use rkyv::{archived_root, Archive, Deserialize, Infallible, Serialize};
+use rusk_uplink::ContractId;
+use std::fs;
+use std::io;
+use std::io::ErrorKind;
+use std::path::Path;
+
+use crate::contract::Contract;
+use crate::state::{Contracts, NetworkState};
+use crate::VMError;
+
+/// The [`NetworkStateId`] is the persisted id of the [`NetworkState`]
+#[derive(Archive, Serialize, Deserialize, Default, Clone, Debug)]
+pub struct NetworkStateId {
+    origin: OffsetLen,
+    head: OffsetLen,
+}
+
+impl NetworkStateId {
+    /// Read from the given path a [`NetworkStateId`]
+    pub fn read<P: AsRef<Path>>(path: P) -> Result<Self, VMError> {
+        let buf = fs::read(&path)?;
+        let id = unsafe { archived_root::<NetworkStateId>(buf.as_slice()) };
+        let id: NetworkStateId = id.deserialize(&mut Infallible).unwrap();
+        Ok(id)
+    }
+
+    /// Write to the given path a [`NetworkStateId`]
+    pub fn write<P: AsRef<Path>>(&self, path: P) -> Result<(), VMError> {
+        let mut serializer = AllocSerializer::<0>::default();
+        serializer.serialize_value(self).unwrap();
+        let bytes = serializer.into_serializer().into_inner();
+        fs::write(&path, bytes.as_slice())?;
+        Ok(())
+    }
+}
+
+impl NetworkState {
+    const PERSISTENCE_ID_FILE_NAME: &'static str = "persist_id";
+
+    /// Persists the origin contracts stored on the [`NetworkState`]
+    pub fn persist(
+        &self,
+        store: StoreRef<OffsetLen>,
+    ) -> Result<NetworkStateId, io::Error> {
+        let head_stored = store.store(&self.head.0);
+        let origin_stored = store.store(&self.origin.0);
+        store.persist().map_err(|_| {
+            io::Error::new(
+                ErrorKind::Other,
+                VMError::PersistenceError(String::from("network state")),
+            )
+        })?;
+        Ok(NetworkStateId {
+            head: *head_stored.ident().erase(),
+            origin: *origin_stored.ident().erase(),
+        })
+    }
+
+    /// Persists the state to disk along with persistence id
+    pub fn persist_to_disk<P: AsRef<Path>>(
+        &self,
+        store: StoreRef<OffsetLen>,
+        store_path: P,
+    ) -> Result<NetworkStateId, VMError> {
+        let persistence_id = self.persist(store).expect("Error in persistence");
+
+        let file_path =
+            store_path.as_ref().join(Self::PERSISTENCE_ID_FILE_NAME);
+
+        persistence_id.write(file_path)?;
+
+        Ok(persistence_id)
+    }
+
+    /// Given a [`NetworkStateId`] restores both [`Hamt`] which store
+    /// contracts of the entire blockchain state.
+    pub fn restore(
```

Comment:
Should this not simply return Result<Self> without having to take self as a parameter? This seems un-intutive and non-idiomatic, but i might be missing something.

Reply:
I tend to agree, although this is a holdover from VM1. We might want to do a simple:

```rust
pub fn restore(id: NetworkStateId, store: StoreRef<OffsetLen>) -> Result<Self, io::Error> {
     // (...)
}
```

It is more idiomatic... That said, since this was already like this before, we might want to tackle this in a separate PR. It's up to you @miloszm.

Reply:
Since origin/head/staged management is going to be re-addressed by Rusk, and since this functionality was like that before and is orthogonal to the subject of this PR, I suggest we tackle it in a separate issue, along with relevant Rusk changes